### PR TITLE
feat: login redirect

### LINKED ISSUE #1670: ui: Login redirects
When users visit `https://<host>/<authenticated path>` the UI should redirect them to `https://<host>/login?next=<authenticated path>`

### PR DESCRIPTION
**Pull request by hoyyeva**
## Related Issues
Resolves #1670

### CODE REVIEW

File: ui/components/auth-required.js [hunk 1]
```diff
@@ -4,14 +4,21 @@ import useSWR from 'swr'
 export default function AuthRequired({ children }) {
   const { data: auth, error } = useSWR('/api/users/self')
   const router = useRouter()
+  const { pathname } = router
```
Comment (jmorganca) [inline ui/components/auth-required.js:7]: We need to copy the entire path as a url to conserve query parameters. See `router.asPath`: https://nextjs.org/docs/api-reference/next/router#router-object

File: ui/components/layouts/dashboard.js [hunk 2]
```diff
@@ -23,7 +23,12 @@ function Layout({ children }) {
       method: 'POST',
     })
     await mutate('/api/users/self', async () => undefined)
-    router.replace('/login')
+
+    window.localStorage.setItem('logout', true)
```
Comment (jmorganca) [inline ui/components/layouts/dashboard.js:27]: Storing this temporary state seems like a band-aid for logic somewhere else – what issue is this fixing – perhaps we can fix that in the `AuthRequired` component

File: ui/components/layouts/dashboard.js [hunk 3]
```diff
@@ -45,13 +50,13 @@ function Layout({ children }) {
   // redirect non-admin routes if user isn't admin
   if (router.pathname.startsWith('/settings') && !accessToSettingsPage) {
     router.replace('/')
-    return null
+    return false
```
Comment (jmorganca) [inline ui/components/layouts/dashboard.js:53]: I think since this is a component it's more idiomatic to return `null`

**Comment by jmorganca** [comment]
Great start! Added few comments

File: ui/components/layouts/dashboard.js (hunk 2 above)
Comment (BruceMacD) [reply ui/components/layouts/dashboard.js:27]: If this is for something like #2526 it should be fixed now

Comment (hoyyeva) [reply ui/components/layouts/dashboard.js:27]: the reason why we store the temporary state in the dashboard is that we won't be able to differentiate if we click logout or not from the previous page.

File: ui/pages/login/callback.js [hunk 4]
```diff
@@ -29,8 +33,10 @@ export default function Callback() {
     const providerID = window.localStorage.getItem('providerID')
     const redirectURL = window.localStorage.getItem('redirectURL')
 
+    const next = window.localStorage.getItem('next')
+
     if (!params.code || !providerID || !redirectURL) {
-      router.replace('/login')
+      next ? router.replace(`/login?next=${next}`) : router.replace('/login')
```
Comment (jmorganca) [inline ui/pages/login/callback.js:39]: Would use an if statement here for clarity

File: ui/pages/groups/index.js [hunk 5]
```diff
@@ -251,7 +251,9 @@ export default function Groups() {
 
   return (
     <>
-      <Head>Groups - Infra</Head>
+      <Head>
```
Comment (jmorganca) [inline ui/pages/groups/index.js:254]: Nice catch!

File: ui/pages/_app.js [hunk 6]
```diff
@@ -29,12 +29,12 @@ function App({ Component, pageProps }) {
   const router = useRouter()
 
   if (!signup) {
-    return null
+    return false
```
Comment (jmorganca) [inline ui/pages/_app.js:32]: Returning `null` is okay here since this is a component

File: ui/components/auth-required.js [hunk 7]
```diff
@@ -4,14 +4,21 @@ import useSWR from 'swr'
...
 
   if (!auth && !error) {
     return null
   }
 
   if (!auth?.id) {
-    router.replace('/login')
-    return null
+    if (!logout && asPath !== '/destinations' && asPath !== '/') {
+      router.replace(`/login?next=${asPath.slice(1)}`)
```
Comment (jmorganca) [inline ui/components/auth-required.js:16]: Do we need to slice here?

Also, I think we'd need to escape `asPath`, e.g:

```router.replace(`/login?next=${encodeURIComponent(asPath)}`)```

Comment (hoyyeva) [reply ui/components/auth-required.js:16]: yes, we need to slice to remove the `/`

File: ui/components/auth-required.js [hunk 8]
```diff
@@ -4,14 +4,21 @@ import useSWR from 'swr'
...
 
   if (!auth && !error) {
     return null
   }
 
   if (!auth?.id) {
-    router.replace('/login')
-    return null
+    if (!logout && asPath !== '/destinations' && asPath !== '/') {
+      router.replace(`/login?next=${encodeURIComponent(asPath.slice(1))}`)
```
Comment (jmorganca) [inline ui/components/auth-required.js:16]: Can remove the `slice(1)` and in below file do `router.replace(next)`

File: ui/pages/_app.js [hunk 9]
```diff
@@ -34,7 +34,7 @@ function App({ Component, pageProps }) {
 
   if (signup.enabled && router.pathname !== '/signup') {
     router.replace('/signup')
-    return null
+    return false
```
Comment (jmorganca) [inline ui/pages/_app.js:37]: `return null` ?

File: ui/pages/index.js [hunk 10]
```diff
@@ -2,6 +2,5 @@ import { useRouter } from 'next/router'
 
 export default function Index() {
   useRouter().replace('/destinations')
-
-  return null
+  return false
```
Comment (jmorganca) [inline ui/pages/index.js:5]: `return null`

File: ui/components/auth-required.js (hunk 8 above)
Comment (jmorganca) [inline ui/components/auth-required.js:16]:
```suggestion
      router.replace(`/login?next=${encodeURIComponent(asPath)}`)
```

File: ui/components/layouts/login.js [hunk 11]
```diff
@@ -4,18 +4,22 @@ import { useRouter } from 'next/router'
...
   if (!auth && !error) {
     return null
   }
 
   if (auth?.id) {
-    // TODO (https://github.com/infrahq/infra/issues/1441): remove me when
-    // using an OTP doesn't trigger authentication
-    if (router.pathname !== '/login/finish') {
+    if (next) {
+      router.replace(`/${next}`)
```
Comment (jmorganca) [inline ui/components/layouts/login.js:15]:
```suggestion
      router.replace(decodeURIComponent(next))
```

**Pull request closed without merging**